src/disjoint_mut.rs

```diff
@@ -0,0 +1,615 @@
+                .expect("Expected range {range:?} to be in the active ranges");
+            ranges.remove(idx);
+        }
+    }
+
+    pub struct DisjointMutGuard<'a, T: AsMutPtr, V: ?Sized> {
+        parent: &'a DisjointMut<T>,
+        slice: &'a mut V,
+        range: (Bound<usize>, Bound<usize>),
+    }
+
+    impl<'a, T: AsMutPtr, V: ?Sized> Drop for DisjointMutGuard<'a, T, V> {
+        fn drop(&mut self) {
+            self.parent.remove_range(self.range, true);
+        }
+    }
+
+    impl<'a, T: AsMutPtr, V: ?Sized> Deref for DisjointMutGuard<'a, T, V> {
+        type Target = V;
+
+        fn deref(&self) -> &Self::Target {
+            self.slice
+        }
+    }
+
+    impl<'a, T: AsMutPtr, V: ?Sized> DerefMut for DisjointMutGuard<'a, T, V> {
+        fn deref_mut(&mut self) -> &mut Self::Target {
+            self.slice
+        }
+    }
+
+    pub struct DisjointImmutGuard<'a, T: AsMutPtr, V: ?Sized> {
+        parent: &'a DisjointMut<T>,
+        slice: &'a V,
+        range: (Bound<usize>, Bound<usize>),
+    }
+
+    impl<'a, T: AsMutPtr, V: ?Sized> Drop for DisjointImmutGuard<'a, T, V> {
+        fn drop(&mut self) {
+            self.parent.remove_range(self.range, false);
+        }
+    }
+
+    impl<'a, T: AsMutPtr, V: ?Sized> Deref for DisjointImmutGuard<'a, T, V> {
+        type Target = V;
+
+        fn deref(&self) -> &Self::Target {
+            self.slice
+        }
+    }
+}
+
+impl<V: Clone> DisjointMut<Vec<V>> {
+    #[allow(unused)]
+    pub fn resize(&mut self, new_len: usize, value: V) {
+        self.inner.get_mut().resize(new_len, value)
+    }
+}
+
+impl<V> DisjointMut<Vec<V>> {
+    #[allow(unused)]
+    pub fn clear(&mut self) {
+        self.inner.get_mut().clear()
+    }
+
+    #[allow(unused)]
+    pub fn resize_with<F>(&mut self, new_len: usize, f: F)
+    where
+        F: FnMut() -> V,
+    {
+        self.inner.get_mut().resize_with(new_len, f)
+    }
+}
+
+unsafe impl<V> AsMutPtr for Vec<V> {
+    type Target = V;
+
+    unsafe fn as_mut_slice(ptr: *mut Self) -> *mut [Self::Target] {
+        let len = unsafe { (*ptr.cast_const()).len() };
+        // SAFETY: .as_mut_ptr() does not materialize a mutable reference to the
+        // underlying slice so we can still allow immutable references into this
+        // slice.
+        let data = unsafe { (*ptr).as_mut_ptr() };
+        ptr::slice_from_raw_parts_mut(data, len)
+    }
+
+    fn len(&self) -> usize {
+        self.len()
+    }
+}
+
+unsafe impl<V, const N: usize> AsMutPtr for [V; N] {
+    type Target = V;
+
+    unsafe fn as_mut_slice(ptr: *mut Self) -> *mut [V] {
+        ptr::slice_from_raw_parts_mut(ptr as *mut V, N)
+    }
+
+    fn len(&self) -> usize {
+        N
+    }
+}
+
+impl<V, const A: usize, const B: usize> DisjointMut<[[V; A]; B]> {
+    // TODO(SJC): Add debug ranges for this method or make this return a raw pointer?
+    #[allow(unused)]
+    pub unsafe fn nd_index_mut(&self, index: [usize; 2]) -> &mut V {
```

Comment:
Are we expecting to add versions of this for more deeply nested arrays? If so maybe name this `index_mut_2d` so that we can more easily add 3d or 4d versions as needed.

Reply:
So if we use `nd_index_mut` we're not getting any disjointness checking, right? What's the reason for this function then? Based on the TODO is the plan to add debug checks here?

Reply:
Correct, I was planning to add a 2d range debug thing, just haven't yet.

Reply:
Were you planning on adding the 2d debug range as part of this PR, or was that meant to be a separate PR? I worry that if we merge this without that we're going to end up with a hole in the `DisjointMut` coverage any place that we try to use `nd_index_mut`, but I'm not sure if you have a reason to think that won't be an issue in practice.

Reply:
> Are we expecting to add versions of this for more deeply nested arrays? If so maybe name this `index_mut_2d` so that we can more easily add 3d or 4d versions as needed.

I agree, this one's only for 2d arrays, right?

Reply:
I thought for a bit on how to do this properly, and then realized the only place we use it could be atomics instead... So I just did that and removed this function entirely. It was always a bit out of place.